### PR TITLE
fix(timezone): Change the time zone in the api data to UTC

### LINKED ISSUE #102316: Inconsistent timezone shown for the same object
#### What happened:
When you query an object, say `Node` object the timestamp will show in local timezone, for example
`kubectl describe node ip-10-0-49-125.us-east-2.compute.internal`
```
CreationTimestamp:  Sun, 16 May 2021 23:26:04 -0400
Taints:             node-role.kubernetes.io/master:NoSchedule
Unschedulable:      false
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  MemoryPressure   False   Mon, 17 May 2021 06:11:57 -0400   Sun, 16 May 2021 23:26:04 -0400   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Mon, 17 May 2021 06:11:57 -0400   Sun, 16 May 2021 23:26:04 -0400   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Mon, 17 May 2021 06:11:57 -0400   Sun, 16 May 2021 23:26:04 -0400   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Mon, 17 May 2021 06:11:57 -0400   Sun, 16 May 2021 23:29:54 -0400   KubeletReady                 kubelet is posting ready status
```
But when you query the same object using `kubectl get node ip-10-0-53-20.us-east-2.compute.internal -o yaml` he same timestamp item will show in UTC,
```
creationTimestamp: "2021-05-17T03:37:49Z"
  conditions:
  - lastHeartbeatTime: "2021-05-18T03:38:18Z"
    lastTransitionTime: "2021-05-17T03:37:49Z"
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: "2021-05-18T03:38:18Z"
    lastTransitionTime: "2021-05-17T03:37:49Z"
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: "2021-05-18T03:38:18Z"
    lastTransitionTime: "2021-05-17T03:37:49Z"
    message: kubelet has sufficient PID available
    reason: KubeletHasSufficientPID
    status: "False"
    type: PIDPressure
  - lastHeartbeatTime: "2021-05-18T03:38:18Z"
    lastTransitionTime: "2021-05-17T03:39:00Z"
    message: kubelet is posting ready status
    reason: KubeletReady
    status: "True"
    type: Ready
```

#### What you expected to happen:
Shouldn't the timezone shown should be consistent across different queries?

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by cndoit18**
Signed-off-by: cndoit18 <cndoit18@outlook.com>

#### What type of PR is this?

/kind bug
/sig node

#### What this PR does / why we need it:
Because of the implementation logic of time.Format in golang, the displayed time zone is not consistent

https://github.com/golang/go/blob/3075ffc93e962792ddf43b2a528ef19b1577ffb7/src/time/format.go#L576

This may cause inconsistencies in the time zones of the data stored on different nodes
 
#### Does this PR introduce a user-facing change?

```release-note
Because of the implementation logic of time.Format in golang, the displayed time zone is not consistent
```

### CODE REVIEW

**Comment by cndoit18** [comment]
I'm not sure if they should prefer to use UTC or the local time zone.
Do you have time to join the discussion? thank you
cc @liggitt

**Comment by neolit123** [comment]
some of the changes here seem dangerous and can create unexpected results when objects already exist in the cluster and there is a sudden change in controller behavior.

less dangerous changes seem to be the /test/* logging timestamps and perhaps the kubectl output too.

**Comment by cndoit18** [comment]
> some of the changes here seem dangerous and can create unexpected results when objects already exist in the cluster and there is a sudden change in controller behavior.
> 
> less dangerous changes seem to be the /test/* logging timestamps and perhaps the kubectl output too.

You are right, we should define a scope. Then I can try to do it from kubectl only

**Comment by cndoit18** [comment]
> some of the changes here seem dangerous and can create unexpected results when objects already exist in the cluster and there is a sudden change in controller behavior.
> 
> less dangerous changes seem to be the /test/* logging timestamps and perhaps the kubectl output too.

If we simply compare two time strings to determine if the time is the same, it may have problems

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go [hunk 1]
```diff
@@ -752,7 +752,7 @@ func describePod(pod *corev1.Pod, events *corev1.EventList) (string, error) {
 			w.Write(LEVEL_0, "Node:\t%s\n", pod.Spec.NodeName+"/"+pod.Status.HostIP)
```
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:752]: keeping describe outputting human-readable times in the current timezone seems better to me

File: staging/src/k8s.io/client-go/util/certificate/certificate_store.go [hunk 2]
```diff
@@ -185,7 +185,7 @@ func loadFile(pairFile string) (*tls.Certificate, error) {
 }
 
 func (s *fileStore) Update(certData, keyData []byte) (*tls.Certificate, error) {
-	ts := time.Now().Format("2006-01-02-15-04-05")
+	ts := time.Now().UTC().Format("2006-01-02-15-04-05")
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/certificate/certificate_store.go:188]: this is used for a local file on disk, no need to convert to utc

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 3]
```diff
@@ -249,7 +249,7 @@ func (t *cachedTokenSource) update(tok *oauth2.Token) map[string]string {
 		ret[k] = v
 	}
 	ret["access-token"] = t.accessToken
-	ret["expiry"] = t.expiry.Format(time.RFC3339Nano)
+	ret["expiry"] = t.expiry.UTC().Format(time.RFC3339Nano)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go:252]: this is only used locally on disk, no need to convert to utc

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 4]
```diff
@@ -484,7 +484,7 @@ func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64, descr1, descr2 inte
 			bestQueueIdx, bestQueueLen = queueIdx, thisLen
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:484]: revert these changes, it is ok for logs to have local timestamps

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 5]
```diff
@@ -342,7 +342,7 @@ func (cfgCtlr *configController) processNextWorkItem() bool {
 // local configController accordingly.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:342]: revert these changes, it is ok for logs to have local timestamps

File: pkg/registry/core/pod/strategy.go [hunk 6]
```diff
@@ -436,7 +436,7 @@ func LogLocation(
 		params.Add("sinceSeconds", strconv.FormatInt(*opts.SinceSeconds, 10))
 	}
 	if opts.SinceTime != nil {
-		params.Add("sinceTime", opts.SinceTime.Format(time.RFC3339))
+		params.Add("sinceTime", opts.SinceTime.UTC().Format(time.RFC3339))
```
Comment (liggitt) [inline pkg/registry/core/pod/strategy.go:439]: this is only used as a query parameter, it isn't important for it to be in UTC

File: pkg/kubelet/logs/container_log_manager.go [hunk 7]
```diff
@@ -400,7 +400,7 @@ func (c *containerLogManager) compressLog(log string) error {
 // rotateLatestLog rotates latest log without compression, so that container can still write
 // and fluentd can finish reading.
 func (c *containerLogManager) rotateLatestLog(id, log string) error {
-	timestamp := c.clock.Now().Format(timestampFormat)
+	timestamp := c.clock.Now().UTC().Format(timestampFormat)
```
Comment (liggitt) [inline pkg/kubelet/logs/container_log_manager.go:403]: this is only used as a local file, it isn't important for it to be in UTC

File: pkg/controller/cronjob/cronjob_controller.go [hunk 8]
```diff
@@ -285,7 +285,7 @@ func syncOne(cj *batchv1.CronJob, js []batchv1.Job, now time.Time, jc jobControl
 	}
 	if tooLate {
 		klog.V(4).Infof("Missed starting window for %s", nameForLog)
-		recorder.Eventf(cj, v1.EventTypeWarning, "MissSchedule", "Missed scheduled time to start a job: %s", scheduledTime.Format(time.RFC1123Z))
+		recorder.Eventf(cj, v1.EventTypeWarning, "MissSchedule", "Missed scheduled time to start a job: %s", scheduledTime.UTC().Format(time.RFC1123Z))
```
Comment (liggitt) [inline pkg/controller/cronjob/cronjob_controller.go:288]: where does the timezone in `scheduledTime` come from? is it the controller time zone?

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 9]
```diff
@@ -501,7 +501,7 @@ func (jm *ControllerV2) syncCronJob(
 	}
 	if tooLate {
 		klog.V(4).InfoS("Missed starting window", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
-		jm.recorder.Eventf(cj, corev1.EventTypeWarning, "MissSchedule", "Missed scheduled time to start a job: %s", scheduledTime.Format(time.RFC1123Z))
+		jm.recorder.Eventf(cj, corev1.EventTypeWarning, "MissSchedule", "Missed scheduled time to start a job: %s", scheduledTime.UTC().Format(time.RFC1123Z))
```
Comment (liggitt) [inline pkg/controller/cronjob/cronjob_controllerv2.go:504]: same question about the source of the timezone in scheduledTime

File: pkg/controller/endpoint/endpoints_controller.go [hunk 10]
```diff
@@ -529,7 +529,7 @@ func (e *Controller) syncService(key string) error {
 
 	if !endpointsLastChangeTriggerTime.IsZero() {
 		newEndpoints.Annotations[v1.EndpointsLastChangeTriggerTime] =
-			endpointsLastChangeTriggerTime.Format(time.RFC3339Nano)
+			endpointsLastChangeTriggerTime.UTC().Format(time.RFC3339Nano)
```
Comment (liggitt) [inline pkg/controller/endpoint/endpoints_controller.go:532]: cc @robscott - is changing this annotation value to UTC problematic?

File: pkg/controller/endpointslice/utils.go [hunk 11]
```diff
@@ -223,7 +223,7 @@ func addTriggerTimeAnnotation(endpointSlice *discovery.EndpointSlice, triggerTim
 	}
 
 	if !triggerTime.IsZero() {
-		endpointSlice.Annotations[corev1.EndpointsLastChangeTriggerTime] = triggerTime.Format(time.RFC3339Nano)
+		endpointSlice.Annotations[corev1.EndpointsLastChangeTriggerTime] = triggerTime.UTC().Format(time.RFC3339Nano)
```
Comment (liggitt) [inline pkg/controller/endpointslice/utils.go:226]: cc @robscott - same question here

File: pkg/controller/podautoscaler/metrics/legacy_metrics_client.go [hunk 12]
```diff
@@ -141,7 +141,7 @@ func (h *HeapsterMetricsClient) GetRawMetric(metricName string, namespace string
 		metricName)
 
 	resultRaw, err := h.services.
-		ProxyGet(h.heapsterScheme, h.heapsterService, h.heapsterPort, metricPath, map[string]string{"start": startTime.Format(time.RFC3339)}).
+		ProxyGet(h.heapsterScheme, h.heapsterService, h.heapsterPort, metricPath, map[string]string{"start": startTime.UTC().Format(time.RFC3339)}).
```
Comment (liggitt) [inline pkg/controller/podautoscaler/metrics/legacy_metrics_client.go:144]: revert this, the query parameter doesn't need to change to UTC

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 13]
```diff
@@ -241,7 +241,7 @@ func (w *logWriter) write(msg *logMessage) error {
 	}
 	line := msg.log
 	if w.opts.timestamp {
-		prefix := append([]byte(msg.timestamp.Format(timeFormatOut)), delimiter[0])
+		prefix := append([]byte(msg.timestamp.UTC().Format(timeFormatOut)), delimiter[0])
```
Comment (liggitt) [inline pkg/kubelet/kuberuntime/logs/logs.go:244]: if other log messages use local time, this probably should remain local time as well

File: staging/src/k8s.io/apiserver/pkg/audit/format.go [hunk 14]
```diff
@@ -61,7 +61,7 @@ func EventString(ev *auditinternal.Event) string {
 	}
 
 	return fmt.Sprintf("%s AUDIT: id=%q stage=%q ip=%q method=%q user=%q groups=%q as=%q asgroups=%q user-agent=%q namespace=%q uri=%q response=\"%s\"",
-		ev.RequestReceivedTimestamp.Format(time.RFC3339Nano), ev.AuditID, ev.Stage, ip, ev.Verb, username, groups, asuser, asgroups, ev.UserAgent, namespace, ev.RequestURI, response)
+		ev.RequestReceivedTimestamp.UTC().Format(time.RFC3339Nano), ev.AuditID, ev.Stage, ip, ev.Verb, username, groups, asuser, asgroups, ev.UserAgent, namespace, ev.RequestURI, response)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/audit/format.go:64]: how is this method used? if it is in a log file that is also using local timestamps, local may be better than utc

File: pkg/kubelet/dockershim/libdocker/fake_client.go [hunk 15]
```diff
@@ -798,7 +798,7 @@ func createImageFromImageInspect(inspect dockertypes.ImageInspect) *dockertypes.
 
 // dockerTimestampToString converts the timestamp to string
 func dockerTimestampToString(t time.Time) string {
-	return t.Format(time.RFC3339Nano)
+	return t.UTC().Format(time.RFC3339Nano)
```
Comment (liggitt) [inline pkg/kubelet/dockershim/libdocker/fake_client.go:801]: how is this method used? what are the implications of changing from local to utc here?

File: pkg/kubelet/types/types.go [hunk 16]
```diff
@@ -58,7 +58,7 @@ func (t *Timestamp) Get() time.Time {
 // GetString returns the time in the string format using the RFC3339NanoFixed
 // layout.
 func (t *Timestamp) GetString() string {
-	return t.time.Format(RFC3339NanoFixed)
+	return t.time.UTC().Format(RFC3339NanoFixed)
```
Comment (liggitt) [inline pkg/kubelet/types/types.go:61]: how is this method used? what are the implications of changing from local to utc here?

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go [hunk 17]
```diff
@@ -38,7 +38,7 @@ func defaultObjectRestarter(obj runtime.Object) ([]byte, error) {
 		if obj.Spec.Template.ObjectMeta.Annotations == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:38]: since this is setting annotations essentially as an API field, forming these in UTC could be worthwhile... how are these consumed/used?

**Comment by cndoit18** [comment]
https://github.com/kubernetes/kubernetes/pull/102451

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go [hunk 18]
```diff
@@ -90,7 +90,7 @@ func Unix(sec int64, nsec int64) Time {
 
 // Rfc3339Copy returns a copy of the Time at second-level precision.
 func (t Time) Rfc3339Copy() Time {
-	copied, _ := time.Parse(time.RFC3339, t.Format(time.RFC3339))
+	copied, _ := time.Parse(time.RFC3339, t.UTC().Format(time.RFC3339))
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go:93]: there are a lot of ways of constructing a new Time instance... it seems odd to make this one coerce to UTC and leave NewTime/Date/Now/Unix/UnmarshalJSON/UnmarshalQueryParameter returning local times

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go:93]: as far as wire compatibility, MarshalJSON/ToUnstructured/MarshalQueryParameter already coerce to UTC, so I tend to think that making the New/Unmarshal methods do the same is *probably* a good idea, but I could also imagine that changing in-memory behavior or failing deep-equal comparisons with locally constructed times

Comment (cndoit18) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go:93]: > as far as wire compatibility, MarshalJSON/ToUnstructured/MarshalQueryParameter already coerce to UTC, so I tend to think that making the New/Unmarshal methods do the same is _probably_ a good idea, but I could also imagine that changing in-memory behavior or failing deep-equal comparisons with locally constructed times

You are right, this UTC is meaningless under time.Parse

File: pkg/controller/endpoint/endpoints_controller.go (hunk 10 above)
Comment (robscott) [reply pkg/controller/endpoint/endpoints_controller.go:532]: Thanks for catching this. I think it would result in some potentially weird values for the SyncProxyRulesLatenct kube-proxy metric after an upgrade if the master wasn't already running in UTC. cc @wojtek-t who added that metric.

File: pkg/controller/endpointslice/utils.go (hunk 11 above)
Comment (robscott) [reply pkg/controller/endpointslice/utils.go:226]: cc @wojtek-t here as well

File: pkg/controller/endpoint/endpoints_controller.go (hunk 10 above)
Comment (wojtek-t) [reply pkg/controller/endpoint/endpoints_controller.go:532]: Hmm...
We parse it here:
https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/endpoints.go#L343
pass it here:
https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/endpoints.go#L226
and then use it here:
https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/iptables/proxier.go#L1632
[by just substracting from now]

I didn't check it, but shouldn't substracting handle UTC time well too?

Comment (cndoit18) [reply pkg/controller/endpoint/endpoints_controller.go:532]: > Hmm...
> We parse it here:
> https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/endpoints.go#L343
> 
> 
> pass it here:
> https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/endpoints.go#L226
> 
> 
> and then use it here:
> https://github.com/kubernetes/kubernetes/blob/3723713c550f649b6ba84964edef9da6cc334f9d/pkg/proxy/iptables/proxier.go#L1632
> 
> 
> [by just substracting from now]
> I didn't check it, but shouldn't substracting handle UTC time well too?

Thank you very much, I looked at it and it is not affected by `time.Format` timezone

File: pkg/controller/cronjob/cronjob_controller.go (hunk 8 above)
Comment (cndoit18) [reply pkg/controller/cronjob/cronjob_controller.go:288]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/cronjob/cronjob_controller.go#L137

Yes, its timezone comes from the controller, and when we don't use the timezone explicitly, the timezone of `time.Now()` will be set to nil, which will make `time.Format()` use the controller's timezone

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 9 above)
Comment (cndoit18) [reply pkg/controller/cronjob/cronjob_controllerv2.go:504]: here as well

File: pkg/kubelet/dockershim/libdocker/fake_client.go (hunk 15 above)
Comment (cndoit18) [reply pkg/kubelet/dockershim/libdocker/fake_client.go:801]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/dockershim/libdocker/fake_client.go#L527

It is only used for fakeclient time generation, it is optional, if we expect the time zone to be UTC, then it should be possible to add it

File: pkg/kubelet/types/types.go (hunk 16 above)
Comment (cndoit18) [reply pkg/kubelet/types/types.go:61]: It is recorded in the annotation of the pod

https://github.com/kubernetes/kubernetes/blob/2f753ec4c826895e4ccd3d6bdda2b1ab777ceeb8/pkg/kubelet/config/config.go#L384

The annotation is used here, parsed by `time.Parse`. So it looks safe
https://github.com/kubernetes/kubernetes/blob/49897ca156c88a3c7a21681bde0b895041db682f/pkg/kubelet/kubelet.go#L1534

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go (hunk 17 above)
Comment (cndoit18) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/kubectl/pkg/cmd/rollout/rollout_restart.go#L159

It is used here to record the data into annotation

File: cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go [hunk 19]
```diff
@@ -61,13 +61,13 @@ func encodeTokenSecretData(token *BootstrapToken, now time.Time) map[string][]by
 	if token.Expires != nil {
 		// Format the expiration date accordingly
 		// TODO: This maybe should be a helper function in bootstraputil?
-		expirationString := token.Expires.Time.Format(time.RFC3339)
+		expirationString := token.Expires.Time.UTC().Format(time.RFC3339)
```
Comment (cndoit18) [inline cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

It will be used here and it looks like the changes should be safe. The scope of the impact is just the Annotation record time for UTC

File: test/e2e/lifecycle/bootstrap/util.go [hunk 20]
```diff
@@ -81,7 +81,7 @@ func addSecretExpiration(s *v1.Secret, expiration string) {
 // TimeStringFromNow returns the time as a string from now.
 // e.g: 2019-12-03T14:30:40+08:00.
 func TimeStringFromNow(delta time.Duration) string {
-	return time.Now().Add(delta).Format(time.RFC3339)
+	return time.Now().Add(delta).UTC().Format(time.RFC3339)
```
Comment (cndoit18) [inline test/e2e/lifecycle/bootstrap/util.go:84]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/test/e2e/lifecycle/bootstrap/bootstrap_token_cleaner.go#L76

It will be used here

**Comment by cndoit18** [comment]
Thank you all very much for the review, I have revised it
cc @liggitt

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go (hunk 17 above)
Comment (liggitt) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: I meant, where is it consumed/displayed

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go [hunk 21]
```diff
@@ -131,7 +132,7 @@ func (s *fsStore) AssignedModified() (time.Time, error) {
 	if err != nil {
 		return time.Time{}, fmt.Errorf("failed to stat %q while checking modification time, error: %v", path, err)
 	}
-	return info.ModTime(), nil
+	return info.ModTime().UTC(), nil
```
Comment (liggitt) [inline pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go:135]: drop this change

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go [hunk 22]
```diff
@@ -300,7 +300,7 @@ func TestFsStoreAssignedModified(t *testing.T) {
 	saveTestSourceFile(t, store, assignedFile, nil)
 
 	// round the current time to the nearest second because some file systems do not support sub-second precision.
-	now := time.Now().Round(time.Second)
+	now := time.Now().UTC().Round(time.Second)
```
Comment (liggitt) [inline pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go:303]: drop this change

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go (hunk 17 above)
Comment (cndoit18) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: > I meant, where is it consumed/displayed

Sorry, I didn't find
It just seems to be recorded and not consumed

截屏2021-06-03 下午10 32 26

File: test/e2e/lifecycle/bootstrap/util.go [hunk 23]
```diff
@@ -81,7 +81,7 @@ func addSecretExpiration(s *v1.Secret, expiration string) {
 // TimeStringFromNow returns the time as a string from now.
 // e.g: 2019-12-03T14:30:40+08:00.
```
Comment (liggitt) [inline test/e2e/lifecycle/bootstrap/util.go:82]: change the godoc to indicate utc

File: pkg/kubelet/types/types.go (hunk 16 above)
Comment (liggitt) [reply pkg/kubelet/types/types.go:61]: This type is used in more places and has more entry points than just this string method. I don't think we should change this as part of this PR. If you want to pursue changing this, split it into its own PR, also address the NewTimestamp and ConvertToTimestamp entry points, resolve all the callers of those entry points to ensure that is safe, and get node/kubelet reviewers to approve

Comment (cndoit18) [reply pkg/kubelet/types/types.go:61]: > This type is used in more places and has more entry points than just this string method. I don't think we should change this as part of this PR. If you want to pursue changing this, split it into its own PR, also address the NewTimestamp and ConvertToTimestamp entry points, resolve all the callers of those entry points to ensure that is safe, and get node/kubelet reviewers to approve

Thank you very much, I understand

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go (hunk 22 above)
Comment (cndoit18) [reply pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go]: Thanks, it's done

File: test/e2e/lifecycle/bootstrap/util.go (hunk 23 above)
Comment (cndoit18) [reply test/e2e/lifecycle/bootstrap/util.go]: Thanks, it's done

File: pkg/kubelet/dockershim/libdocker/fake_client.go (hunk 15 above)
Comment (liggitt) [reply pkg/kubelet/dockershim/libdocker/fake_client.go:801]: if this isn't related to persisting data in the API, drop it from this PR

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go [hunk 24]
```diff
@@ -18,10 +18,11 @@ package store
 
 import (
 	"fmt"
-	"k8s.io/klog/v2"
 	"path/filepath"
 	"time"
 
+	"k8s.io/klog/v2"
```
[the remark this answers is not in the document]
Comment (cndoit18) [reply pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go]: thank you, it is done

File: cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go (hunk 19 above)
Comment (liggitt) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: would like ack from cluster-lifecycle reviewer ...

/assign @neolit123

File: pkg/controller/cronjob/cronjob_controller.go (hunk 8 above)
Comment (liggitt) [reply pkg/controller/cronjob/cronjob_controller.go:288]: would like an ack from a sig-apps review
/assign @soltysh

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 9 above)
Comment (liggitt) [reply pkg/controller/cronjob/cronjob_controllerv2.go:504]: would like an ack from a sig-apps review
/assign @soltysh

File: pkg/controller/endpoint/endpoints_controller.go (hunk 10 above)
Comment (liggitt) [reply pkg/controller/endpoint/endpoints_controller.go:532]: I think this lgtm. @wojtek-t can you explicitly ack this and the other endpoint/endpointslice reconciler changes?

/assign @wojtek-t

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go (hunk 17 above)
Comment (liggitt) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: Ok... that means that it is only displayed as a generic string annotation, and changing it to UTC will take it out of the timezone of the kubectl runner. Is that what we want? That seems like a worse user experience.

Comment (cndoit18) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: If it is stored in annotation, the user can only read it by '-o yaml', which will result in time zone inconsistency in the same yaml. It seems to be more likely to be mistaken for an error, don't you think?

File: cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go (hunk 19 above)
Comment (cndoit18) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: thank you very much

File: staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go (hunk 17 above)
Comment (liggitt) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: I'm not aware of any complaint/confusion about it in the past. The most likely audience to view that is the person that issued the restart command, and preserving their current timezone seems reasonable to me.

Comment (cndoit18) [reply staging/src/k8s.io/kubectl/pkg/polymorphichelpers/objectrestarter.go:39]: I see, maybe we can remember this thing for now and wait until a user has a question and we can restart it again.
Thank you

File: cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go (hunk 19 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: the parsing from cluster-bootstrap seems fine:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

but if generated tokens client side start using UTC, my assumption is that server side the token cleaner and bootstrap signer controllers should be adapted as well.
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/bootstrap/tokencleaner.go#L192

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go#L131

if we make the change server side, to me this seems like we are introducing a weird scenario for tokens that already exist in etcd.

this was mentioned here in the other closed PR:
https://github.com/kubernetes/kubernetes/pull/102451#discussion_r642546852

i could be wrong.

Comment (cndoit18) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: > the parsing from cluster-bootstrap seems fine:
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64
> 
> but if generated tokens client side start using UTC, my assumption is that server side the token cleaner and bootstrap signer controllers should be adapted as well.
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/bootstrap/tokencleaner.go#L192
> 
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go#L131
> 
> if we make the change server side, to me this seems like we are introducing a weird scenario for tokens that already exist in etcd.
> 
> this was mentioned here in the other closed PR:
> [#102451 (comment)](https://github.com/kubernetes/kubernetes/pull/102451#discussion_r642546852)
> 
> i could be wrong.

Thank you very much, do you mean that the time zone used in the secrets already stored in etcd is local, but after the update, the subsequent stored secrets are changed to utc. may cause confusion to the user?

Comment (cndoit18) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

the tokencleaner should not need to be modified, it uses "time.Parse" parsing and will parse to UTC time zone

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: the controllers currently accept local time and `time.Now() != time.Now().UTC()`

Comment (cndoit18) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: > the controllers currently accept local server time and `time.Now() != time.Now().UTC()`

Sorry, I didn't understand

`time.Now().Equal(time.Now().UTC())` is true

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: i think i'm ignoring the fact that when a time object is marshaled to RFC3339 for local time it stores the time zone offset.
which means that parsing the string back to a time object would result in the same time.

so starting to store time as UTC strings in boostrap objects doesn't seem like a breaking change.
it is just cosmetic.

/approve

Comment (cndoit18) [reply cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go:64]: > i think i'm ignoring the fact that when a time object is marshaled to RFC3339 for local time it stores the time zone offset.
> which means that parsing the string back to a time object would result in the same time.
> 
> so starting to store time as UTC strings in boostrap token objects doesn't seem like a breaking change.
> it is just cosmetic.
> 
> /approve

As you said, thanks very much!

**Comment by cndoit18** [comment]
Hi, can someone please help me to execute `ok-to-test`? thank you

**Comment by cndoit18** [comment]
> /ok-to-test

thank you!

File: pkg/controller/endpoint/endpoints_controller.go (hunk 10 above)
Comment (wojtek-t) [reply pkg/controller/endpoint/endpoints_controller.go:532]: I double checked the golang time package code and this change is safe.

So LGTM.

**Review by wojtek-t — COMMENTED** [review]
/approve [for the endpoints part]

**Comment by cndoit18** [comment]
> /approve [for the endpoints part]

thank you very much, I should still need the "sig-apps" approval now

**Comment by cndoit18** [comment]
ping @soltysh 
Can you please help me review it and tag it with LGTM? thank you

File: pkg/controller/cronjob/cronjob_controller.go (hunk 8 above)
Comment (soltysh) [reply pkg/controller/cronjob/cronjob_controller.go:288]: Yeah, this should be fine.

**Review by soltysh — APPROVED** [review]
/lgtm
sig-apps changes
